### PR TITLE
FAI-943: Add ability for TrustyAI to call KServe/ModelMesh models directly

### PR DESCRIPTION
**Pull request by ruivieira**
[FAI-943](https://issues.redhat.com/browse/FAI-943).

Initial implementation of a KServe/ModelMesh `PredictionProvider` which allows TrustyAI explainers and metrics to call a model server directly.

This PR adds a gRPC-based `PredictionProvider` that can be used directly in TrustyAI.
The gRPC client is based on the [KServe v2 protobuf](https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/grpc_predict_v2.proto).

In this implementation only the model server endpoint, model name (and optionally output feature names) need to be specified. As an example for an exposed endpoint at `localhost:8033` we could use it with LIME as

```java
 final PredictionProvider kserveProvider = KServeV2GRPCPredictionProvider
                .forTarget("localhost:8033", "example-sklearn-isvc");

List<PredictionInput> input = ...

final PredictionOutput outputs = kserveProvider.predictAsync(input).get().get(0);
final Prediction prediction = new SimplePrediction(positiveInput, outputs);
final LimeExplainer lime = new LimeExplainer();
final SaliencyResults explanation = lime.explainAsync(prediction, kserveProvider).get();
```

gRPC client can be re-generated using `mvn clean compile`.

> Many thanks for submitting your Pull Request :heart:!
> 
> Please make sure that your PR meets the following requirements:
> 
> - [x] You have read the [contributors guide](https://github.com/trustyai-explainability/trustyai-explainability/blob/main/CONTRIBUTING.md)
> - [x] Pull Request title is properly formatted: `FAI-XYZ Subject`
> - [x] Pull Request title contains the target branch if not targeting main: `[0.3.x] FAI-XYZ Subject`
> - [x] Pull Request contains link to the JIRA issue
> - [x] Pull Request contains link to any dependent or related Pull Request
> - [x] Pull Request contains description of the issue
> - [x] Pull Request does not include fixes for issues other than the main ticket
> 
>

### CODE REVIEW

File: explainability-connectors/pom.xml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        <maven.compiler.target>11</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <grpc.version>1.52.1</grpc.version><!-- CURRENT_GRPC_VERSION -->
+        <protoc.version>3.19.3</protoc.version>
+    </properties>
+    <dependencies>
+        <dependency>
+            <groupId>org.kie.trustyai</groupId>
+            <artifactId>explainability-core</artifactId>
+            <version>999-SNAPSHOT</version>
```
Comment (danielezonca) [inline explainability-connectors/pom.xml]:
```suggestion
            <version>${project.version}</version>
```

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    public static KServeV2GRPCPredictionProvider forTarget(String target, String modelName, List<String> outputNames) {
+        return new KServeV2GRPCPredictionProvider(target, modelName, outputNames);
+    }
+
+    private void addFeature(InferTensorContents.Builder content, Feature feature) {
+        final Object object = feature.getValue().getUnderlyingObject();
+        final Type type = feature.getType();
+
+        switch (type) {
+            case NUMBER:
```
Comment (RobGeada) [inline explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java:71]: Can the GRPC data only be numeric or boolean, or is this just a MVP implementation? If the former, we should check for + allow numeric categoricals as well.

File: explainability-connectors/src/main/proto/grpc_predict_v2.proto [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+// limitations under the License.
+
+syntax = "proto3";
+package inference;
+
+option java_multiple_files = true;
+option java_package = "org.kie.trustyai.connectors.kserve.v2.grpc";
+
+// Inference Server GRPC endpoints.
+service GRPCInferenceService
```
Comment (danielezonca) [inline explainability-connectors/src/main/proto/grpc_predict_v2.proto:22]: I guess this is a file copied from kserve project. Can you add this info in the readme? I.e. the version used/how to update/url

File: explainability-connectors/pom.xml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-protobuf</artifactId>
+            <version>${grpc.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>jakarta.annotation</groupId>
+            <artifactId>jakarta.annotation-api</artifactId>
```
Comment (danielezonca) [inline explainability-connectors/pom.xml:36]: Where is it used?

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/utils/ListenableFutureUtils.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    private final CompletableFuture<T> completableFuture;
+
+    public ListenableFutureUtils(ListenableFuture<T> listenableFuture) {
+        this.completableFuture = new CompletableFuture<T>() {
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning) {
+                boolean cancelled = listenableFuture.cancel(mayInterruptIfRunning);
+                super.cancel(cancelled);
+                return cancelled;
```
Comment (danielezonca) [inline explainability-connectors/src/main/java/org/kie/trustyai/connectors/utils/ListenableFutureUtils.java:25]: I'm not a fan of "extreme single lining" but isn't this equivalent to
```suggestion
                return super.cancel(listenableFuture.cancel(mayInterruptIfRunning));
```
?

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+     * @param target The remote KServe v2 model server
+     * @param modelName The model's name
+     * @param outputNames A {@link List} of output names to be used
+     * @return A {@link PredictionProvider}
+     */
+    public static KServeV2GRPCPredictionProvider forTarget(String target, String modelName, List<String> outputNames) {
+        return new KServeV2GRPCPredictionProvider(target, modelName, outputNames);
+    }
+
+    private void addFeature(InferTensorContents.Builder content, Feature feature) {
```
Comment (danielezonca) [inline explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java:66]: I think it is fine to start with number/boolean only but it should be nice to expand it to cover more (most?) of the supported `Type` of TrustyAI

**Review by danielezonca — COMMENTED** [review]
Few comments but overall very well done :)

**Review by RobGeada — COMMENTED** [review]
This is really great, thanks for this work. Just got a comment on Type inference, but otherwise looks good to me.

File: explainability-connectors/pom.xml (hunk 4 above)
Comment (ruivieira) [reply explainability-connectors/pom.xml:36]: @danielezonca AFAICT this is needed for the gRPC client code that uses `@javax.annotation.Generated`. Is there an alternative?

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java (hunk 6 above)
Comment (ruivieira) [reply explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java:66]: I agree. We could merge this PR in order to test with Python first and then add missing feature types on a separate PR. wdyt?

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java (hunk 2 above)
Comment (ruivieira) [reply explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java:71]: @RobGeada The idea was to merge with a few types, so we can also test with Python and add remaining feature types next. wdyt? There are more types supported, for instance strings/bytes.

**Review by tteofili — APPROVED** [review]
thanks @ruivieira LGTM!

File: explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java (hunk 6 above)
Comment (danielezonca) [reply explainability-connectors/src/main/java/org/kie/trustyai/connectors/kserve/v2/KServeV2GRPCPredictionProvider.java:66]: Fine for me 👍

File: explainability-connectors/pom.xml (hunk 4 above)
Comment (danielezonca) [reply explainability-connectors/pom.xml:36]: It is fine, I just asked because I did not check generated files :)

**Review by danielezonca — APPROVED** [review]

**Pull request merged**